### PR TITLE
Add E2E test for wildcard route interception API route request

### PR DESCRIPTION
**Pull request by hamlim**

### CODE REVIEW

**Comment by hamlim** [comment]
Figured I'd try to implement a fix also, pushed a change here: https://github.com/wakujs/waku/pull/1451/commits/0805f30d47dd8809b4c8327f2ac2db927ea49dc0 to sort wildcard path configs to the end of the array so we match on them last.

Not sure if this is a naive fix (and maybe leads to other issues), but it seems to make the most sense in that we should probably match explicit path specs first before falling back on wildcard route segments.

cc @tylersayshi

**Review by dai-shi — CHANGES_REQUESTED** [review]
> Not sure if this is a naive fix (and maybe leads to other issues), but it seems to make the most sense in that we should probably match explicit path specs first before falling back on wildcard route segments.

It might make sense, but I would like to leave this responsibility for create-pages.ts. Can you modify create-pages.ts instead of define-router.ts?

**Comment by hamlim** [comment]
@dai-shi got some time to take a look at the code again today - I don't think this can be solved for within `create-pages.ts` as the code is today. There's no exposed hook within `defineRouter` that allows `create-pages` to sort the complete list of path specs or to choose the matching path config.

One possible option could be to have `defineRouter` allow for integrations to pass in a custom `getPathConfigItem` function (keep the current implementation as a default parameter, maybe tweak it slightly to call `getPathConfigItem` with the `pathConfig` and the existing `pathname` argument), then `create-pages` could pass in a custom method to sort the list of path configs and then return a `.find` for the path. Would that change make more sense?

After thinking about my naive solve here a bit more, an even simpler solve could be to move the [api route spread here](https://github.com/wakujs/waku/blob/79efb5c6cd48ad03feae40cac3543055a404cd36/packages/waku/src/router/define-router.ts#L233-L243) up above the [page route spread](https://github.com/wakujs/waku/blob/79efb5c6cd48ad03feae40cac3543055a404cd36/packages/waku/src/router/define-router.ts#L204-L232).

**Comment by dai-shi** [comment]
Ah I see what you mean. I think we should make some changes in define-router.ts. I can work on it later but not soon.

**Comment by dai-shi** [comment]
> I can work on it later but not soon.

I haven't started. Give me time.

**Comment by dai-shi** [comment]
> I haven't started. Give me time.

👉 #1487

**Review by dai-shi — CHANGES_REQUESTED** [review]
@hamlim I'm sorry, but you misunderstand my request. The fix should be in create-pages.ts. Do not modify define-router.ts.

**Comment by hamlim** [comment]
Sorry about that, updated the branch to apply the fix in `create-pages` instead!

If `defineRouter` will be more of a stable public API in the long run - we may want to consider documenting this quirk to let folks know that this issue may happen if they're not using the `create-pages` setup - but that can be done in a future change

**Comment by dai-shi** [comment]
Record of flaky tests:

https://github.com/wakujs/waku/actions/runs/15853235995/job/44691999735?pr=1451

> Error:   1) [chromium-prd] › e2e/ssg-performance.spec.ts:17:3 › high volume static site generation › build and verify 
>     Test timeout of 60000ms exceeded.

https://github.com/wakujs/waku/actions/runs/15853235995/job/44691999777?pr=1451

> Error:   1) [chromium-prd] › e2e\ssg-performance.spec.ts:17:3 › high volume static site generation › build and verify 
>     Test timeout of 60000ms exceeded.

**Review by dai-shi — APPROVED** [review]
LGTM

@tylersayshi Can you have another review?

File: packages/waku/src/router/create-pages.ts [hunk 1]
```diff
@@ -753,7 +753,19 @@ export const createPages = <
...
+      return (
+        [...routeConfigs, ...apiConfigs]
+          // Sort wildcard routes to the end - we want to match the most specific routes first
+          .sort((pathConfigA, pathConfigB) => {
+            if (pathConfigA.path.find((spec) => spec.type === 'wildcard')) {
+              return 1;
+            }
+            if (pathConfigB.path.find((spec) => spec.type === 'wildcard')) {
+              return -1;
+            }
```
Comment (tylersayshi) [inline packages/waku/src/router/create-pages.ts]:
```suggestion
            if (pathConfigA.path.at(-1)?.type === 'wildcard') {
              return 1;
            }
            if (pathConfigB.path.at(-1)?.type === 'wildcard') {
              return -1;
            }
```
this would work too, right?

Comment (hamlim) [reply packages/waku/src/router/create-pages.ts]: I realized that there was a possible edge case with the sorting logic (api wildcard routes ending up lower than standard api routes).

I've added another test case and updated the sorting logic - it's a bit more convoluted but I think it mostly makes sense.

fyi @dai-shi

File: packages/waku/src/router/create-pages.ts [hunk 2]
```diff
@@ -753,7 +753,38 @@ export const createPages = <
...
+            let priorityA = 0;
+            if (pathConfigA.type === 'api') {
+              if (hasWildcardA) {
+                priorityA = 2;
+              } else {
+                priorityA = 1;
+              }
+            } else if (hasWildcardA) {
+              priorityA = 3;
+            }
```
Comment (dai-shi) [inline packages/waku/src/router/create-pages.ts]: Make it a function, incl. L760, and reuse the logic. It will be more readable.

Comment (hamlim) [reply packages/waku/src/router/create-pages.ts]: This commit abstracts this logic away into a function: https://github.com/wakujs/waku/pull/1451/commits/a4a6e63a1a754f481d68eccd4026767d91d14f19

File: packages/waku/src/router/create-pages.ts [hunk 3]
```diff
@@ -753,7 +753,27 @@ export const createPages = <
...
+
+      return (
+        [...routeConfigs, ...apiConfigs]
+          // Sort routes by priority: "standard routes" -> api routes -> api wildcard routes -> standard wildcard routes
+          .sort((pathConfigA, pathConfigB) => {
+            const priorityA = getRoutePriority(pathConfigA);
+            const priorityB = getRoutePriority(pathConfigB);
+
+            return priorityA - priorityB;
+          })
```
Comment (dai-shi) [inline packages/waku/src/router/create-pages.ts]: Looks good.

My preference would be:
`sort((configA, configB) => getRoutePriority(configA) - getRoutePriority(configB))`

**Comment by hamlim** [comment]
Thanks for all the reviews and feedback!
